src/jit/inline.cpp

```diff
@@ -474,3 +642,179 @@ void InlineResult::report()
+{
+    // For now, always create a Legacy policy.
+    InlinePolicy* policy = new (compiler, CMK_Inlining) LegacyPolicy;
+
+    return policy;
+}
+
+//------------------------------------------------------------------------
+// noteCandidate: handle passing a set of inlining checks successfully
+//
+// Arguments:
+//    obs      - the current obsevation
+
+void LegacyPolicy::noteCandidate(InlineObservation obs)
+{
+    assert(!inlDecisionIsDecided(inlDecision));
+
+    // Check the impact, it should be INFORMATION
+    InlineImpact impact = inlGetImpact(obs);
+    assert(impact == InlineImpact::INFORMATION);
+
+    // Update the status
+    setCommon(InlineDecision::CANDIDATE, obs);
+}
+
+//------------------------------------------------------------------------
+// noteSuccess: handle finishing all the inlining checks successfully
+
+void LegacyPolicy::noteSuccess()
+{
+    assert(inlDecisionIsCandidate(inlDecision));
+    inlDecision = InlineDecision::SUCCESS;
+}
+
+//------------------------------------------------------------------------
+// note: handle an observation with non-fatal impact
+//
+// Arguments:
+//    obs      - the current obsevation
+
+void LegacyPolicy::note(InlineObservation obs)
+{
+    // Check the impact
+    InlineImpact impact = inlGetImpact(obs);
+
+    // As a safeguard, all fatal impact must be
+    // reported via noteFatal.
+    assert(impact != InlineImpact::FATAL);
+    noteInternal(obs, impact);
+}
+
+//------------------------------------------------------------------------
+// noteFatal: handle an observation with fatal impact
+//
+// Arguments:
+//    obs      - the current obsevation
+
+void LegacyPolicy::noteFatal(InlineObservation obs)
+{
+    // Check the impact
+    InlineImpact impact = inlGetImpact(obs);
+
+    // As a safeguard, all fatal impact must be
+    // reported via noteFatal.
+    assert(impact == InlineImpact::FATAL);
+    noteInternal(obs, impact);
+    assert(inlDecisionIsFailure(inlDecision));
+}
+
+//------------------------------------------------------------------------
+// noteInt: handle an observed integer value
+//
+// Arguments:
+//    obs      - the current obsevation
+//    value    - the value being observed
+
+void LegacyPolicy::noteInt(InlineObservation obs, int value)
+{
+    (void) value;
+    note(obs);
+}
+
+//------------------------------------------------------------------------
+// noteDouble: handle an observed double value
+//
+// Arguments:
+//    obs      - the current obsevation
+//    value    - the value being observed
+
+void LegacyPolicy::noteDouble(InlineObservation obs, double value)
+{
+    (void) value;
+    note(obs);
+}
+
+//------------------------------------------------------------------------
+// setNever: helper for handling an observation
+//
+// Arguments:
+//    obs      - the current obsevation
+//    impact   - impact of the current observation
+
+void LegacyPolicy::noteInternal(InlineObservation obs, InlineImpact impact)
```

Comment:
`noteInternal` is private, the key public callers `note` and `noteFatal` already screen for appropriate impacts. I think this is sufficient for the time being -- legacy policy only really cares about `FATAL` for the most part.